Add new custom derivatives for math functions

a-elrawy opened the pull request:
Contributing to #871 
Added support for the following functions:

- `tan`
- `cosh`
- `sinh`
- `asin`
- `atan`
- `log10`
- `log2`
- `cbrt`
- `hypot`

vgvassilev commented:
This looks quite good. What are the remaining functions from math.h and cmath that still need to be supported?

a-elrawy commented:
These are some funcitons from https://en.cppreference.com/w/c/numeric/math

**Basic Functions:**     
- [x] `abs`, `labs`, `llabs`, `imaxabs`     
- [x] `fabs`, `fabsf`, `fabsl`     
- [ ] `fmod`, `fmodf`, `fmodl`     
- [ ] `remainder`, `remainderf`, `remainderl`     
- [ ] `remquo`, `remquof`, `remquol`     
- [x] `fma`, `fmaf`, `fmal`     
- [ ]  `fmax`, `fmaxf`, `fmaxl`     
- [ ] `fmin`, `fminf`, `fminl`     
- [ ] `fdim`, `fdimf`, `fdiml`

**Exponential Functions:**     
- [x] `exp`, `expf`, `expl`     
- [ ] `exp2`, `exp2f`, `exp2l`     
- [ ] `expm1`, `expm1f`, `expm1l`     
 
**Logarithmic Functions:**    
- [x] `log`, `logf`, `logl`     
- [x] `log10`, `log10f`, `log10l`     
- [x] `log2`, `log2f`, `log2l`     
- [ ] `log1p`, `log1pf`, `log1pl`

**Power Functions:**     
- [x] `pow`, `powf`, `powl`     
- [x] `sqrt`, `sqrtf`, `sqrtl`     
- [x] `cbrt`, `cbrtf`, `cbrtl`
- [x] `hypot`, `hypotf`, `hypotl`

**Trigonometric Functions:**     
- [x] `sin`, `sinf`, `sinl`     
- [x] `cos`, `cosf`, `cosl`    
- [x]  `tan`, `tanf`, `tanl`     
- [x] `asin`, `asinf`, `asinl`     
- [x] `acos`, `acosf`, `acosl`     
- [x] `atan`, `atanf`, `atanl`     
- [x] `atan2`, `atan2f`, `atan2l`  
    
**Hyperbolic Functions:**     
- [x] `sinh`, `sinhf`, `sinhl`    
- [x]  `cosh`, `coshf`, `coshl`     
- [ ] `tanh`, `tanhf`, `tanhl`     
- [ ] `asinh`, `asinhf`, `asinhl`     
- [ ] `acosh`, `acoshf`, `acoshl`     
- [ ] `atanh`, `atanhf`, `atanhl`  
    
**Error and Gamma Functions:**     
- [ ] `erf`, `erff`, `erfl`     
- [ ] `erfc`, `erfcf`, `erfcl`     
- [ ] `tgamma`, `tgammaf`, `tgammal`     
- [ ] `lgamma`, `lgammaf`, `lgammal`

I'm not sure if I should include the remaining functions if they don't have derivatives.
Also, do we've to make a separate custom functions for function like `sin`, `sinf`, `sinl` ?

vgvassilev commented:
> I'm not sure if I should include the remaining functions if they don't have derivatives.

Probably returning 0 is fine. @parth-07 what do you think?

> Also, do we've to make a separate custom functions for function like `sin`, `sinf`, `sinl` ?

Yes. You can still fall back to a single templated implementation function.

parth-07 commented:
> > I'm not sure if I should include the remaining functions if they don't have derivatives.

> Probably returning 0 is fine. @parth-07 what do you think?

@a-elrawy When you say a function does not have a derivative, do you mean the function is non-differentiable due to discontinuity and cusp/corner?

a-elrawy commented:
> > > I'm not sure if I should include the remaining functions if they don't have derivatives.
> 
> > Probably returning 0 is fine. @parth-07 what do you think?
> 
> @a-elrawy When you say a function does not have a derivative, do you mean the function is non-differentiable due to discontinuity and cusp/corner?

I mean functions like floor and ceil (they are implemented to return 0), and other functions after Nearest integer floating-point operations here: https://en.cppreference.com/w/c/numeric/math
 

parth-07 commented:
> I mean functions like floor and ceil (they are implemented to return 0), and other functions after Nearest integer floating-point operations here: https://en.cppreference.com/w/c/numeric/math

Oh, I see. Yes, I think we should return 0 as the derivative for such functions.

parth-07 commented on include/clad/Differentiator/BuiltinDerivatives.h:
```diff
@@ -372,6 +372,104 @@ CUDA_HOST_DEVICE void clamp_pullback(const T& v, const T& lo, const T& hi,
+}
+
+template <typename T, typename dT>
+CUDA_HOST_DEVICE void tan_pullback(T x, T d_y, T* d_x) {
```
`dT` template type is not being used.

a-elrawy replied:
It's updated now

parth-07 commented on include/clad/Differentiator/BuiltinDerivatives.h:
```diff
@@ -372,6 +372,104 @@ CUDA_HOST_DEVICE void clamp_pullback(const T& v, const T& lo, const T& hi,
+template <typename T, typename dT>
+CUDA_HOST_DEVICE ValueAndPushforward<T, dT> tan_pushforward(T x, dT d_x) {
+  T tanx = ::std::tan(x);
+  return {tanx, (1 + tanx * tanx) * d_x};
```
shouldn't the pushforward here be `sec^2(x) * d_x`?

a-elrawy replied:
both should be equivalent. I used this because I think it takes less computations than sec^2.

parth-07 replied:
Oh, yes. You are right.

vgvassilev commented:
@a-elrawy, is there anything else that needs to be done here?

a-elrawy commented:
> @a-elrawy, is there anything else that needs to be done here?

I've added new functions and added the different floating-point implementations. The tests ran successfully on my end, but it's failing for some reason in git actions. Can you please take a look?

vgvassilev commented:
@a-elrawy, this is looking good. Do you want to cover the hyperbolic functions and error and gamma functions to declare complete success here?

a-elrawy commented:
> @a-elrawy, this is looking good. Do you want to cover the hyperbolic functions and error and gamma functions to declare complete success here?

I can add them by Friday.

a-elrawy commented:
> @a-elrawy, this is looking good. Do you want to cover the hyperbolic functions and error and gamma functions to declare complete success here?

The gamma functions may need digamma from boost for calculating the derivative. I've added tgamma in the numericalDiff tests for now

vgvassilev commented:
@a-elrawy, can you squash this into one commit and write a good commit message to automatically close the issue you have fixed?
